### PR TITLE
Implement "CREATE TABLE AS"

### PR DESCRIPTION
**Pull request by jeeminso**
…with column constraints then inserts the subquery results

## Summary of the changes / Why this improves CrateDB
This is regarding #9999 

documentation remaining

 

## Checklist

 - [x] Added an entry in `CHANGES.txt` for user facing changes
 - [x] Updated documentation & `sql_features` table for user facing changes
 - [x] Touched code is covered by tests
 - [x] [CLA](https://crate.io/community/contribute/cla/) is signed
 - [x] This does not contain breaking changes, or if it does:
    - It is released within a major release
    - It is recorded in ``CHANGES.txt``
    - It was marked as deprecated in an earlier release if possible
    - You've thought about the consequences and other components are adapted
      (E.g. AdminUI)

### CODE REVIEW

**Comment by mfussenegger** [comment]
ok to test

File: server/src/main/java/io/crate/analyze/CreateTableAsAnalyzer.java [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+    }
+
+    private List<TableElement<Expression>> getTableElementsFromSourceQuery (Query sourceQuery,
+                                                                            CoordinatorTxnCtx txnCtx) {
+        var querySpec = (QuerySpecification) sourceQuery.getQueryBody();
+        List<Relation> relations = querySpec.getFrom();
+        List<SelectItem> selectItems = querySpec.getSelect().getSelectItems();
+
+        List<TableElement<Expression>> allTableElements = relations.stream()
+            .map(relation -> (Table)relation)
```
Comment (mfussenegger) [inline server/src/main/java/io/crate/analyze/CreateTableAsAnalyzer.java:107]: I think this logic here may break if the query isn't a simple `select ... from tbl` because it assumes there is a `QuerySpecification` and that the relations are `Table` instances.


I think to support all kinds of query it would be possible to use the `RelationAnalyzer` to analyze the query, and then to create the required table elements from the `outputs` of the `AnalyzedRelation`.

**Review by mfussenegger — COMMENTED** [review]
Thank you for the contribution :+1: 

Looks quite good so far. I've left a comment regarding some of the analyzer logic. Other than that to me it looks like it is on the right track.

File: server/src/main/java/io/crate/analyze/CreateTableAsAnalyzer.java (hunk 1 above)
Comment (jeeminso) [reply server/src/main/java/io/crate/analyze/CreateTableAsAnalyzer.java:107]: thanks for reviewing and suggestions! will move on

**Comment by jeeminso** [comment]
I have followed your suggestion and now it handles more sub-types of analyzedRelations and the sub-types of symbols from outputs method. Also added a few test cases. (Currently it does 'new' instead of using variants of 'is', will look into it. )
Once I figure out, i will add more test cases and expand on the extractor logic. ( it probably does not handle all cases yet)
Lastly, please help on the jenkins pr_test issue. Not sure how to handle this myself without looking into the error messages. Thank you.

**Comment by amotl** [comment]
Dear @jeeminso,

thanks a bunch for following up on this. While @mfussenegger will have to take a look at the amendments again, I can share the output of the Jenkins job with you in the meanwhile. It is just the Checkstyle plugin which reported some violations.

### Errors
```
> Task :libs:sql-parser:checkstyleMain

[ant:checkstyle] [ERROR] /var/lib/jenkins/workspace/CrateDB/crate_on_pr/libs/sql-parser/src/main/java/io/crate/sql/tree/AstVisitor.java:283:64: '{' at column 64 should have line break after. [LeftCurly]
[ant:checkstyle] [ERROR] /var/lib/jenkins/workspace/CrateDB/crate_on_pr/libs/sql-parser/src/main/java/io/crate/sql/tree/AstVisitor.java:283:104: '}' at column 104 should be alone on a line. [RightCurlyAlone]
[ant:checkstyle] [ERROR] /var/lib/jenkins/workspace/CrateDB/crate_on_pr/libs/sql-parser/src/main/java/io/crate/sql/tree/CreateTableAs.java:1: Line does not match expected header line of '^/\*'. [RegexpHeader]
[ant:checkstyle] [ERROR] /var/lib/jenkins/workspace/CrateDB/crate_on_pr/libs/sql-parser/src/main/java/io/crate/sql/tree/CreateTableAs.java:5:51: WhitespaceAround: '{' is not preceded with whitespace. [WhitespaceAround]
[ant:checkstyle] [ERROR] /var/lib/jenkins/workspace/CrateDB/crate_on_pr/libs/sql-parser/src/main/java/io/crate/sql/tree/CreateTableAs.java:19:26: '{' at column 26 should have line break after. [LeftCurly]
[ant:checkstyle] [ERROR] /var/lib/jenkins/workspace/CrateDB/crate_on_pr/libs/sql-parser/src/main/java/io/crate/sql/tree/CreateTableAs.java:19:47: '}' at column 47 should be alone on a line. [RightCurlyAlone]

> Task :libs:sql-parser:checkstyleMain FAILED
```


### Warnings
Might not be related to your change.
```
> Task :server:compileTestJava

/var/lib/jenkins/workspace/CrateDB/crate_on_pr/server/src/test/java/org/elasticsearch/test/disruption/LongGCDisruption.java:249: warning: [removal] suspend() in Thread has been deprecated and marked for removal
                        thread.suspend();
                              ^

/var/lib/jenkins/workspace/CrateDB/crate_on_pr/server/src/test/java/org/elasticsearch/test/disruption/LongGCDisruption.java:270: warning: [removal] resume() in Thread has been deprecated and marked for removal
                            thread.resume();
                                  ^

/var/lib/jenkins/workspace/CrateDB/crate_on_pr/server/src/test/java/org/elasticsearch/test/disruption/LongGCDisruption.java:317: warning: [removal] resume() in Thread has been deprecated and marked for removal
            thread.resume();
                  ^
```

You might be able to run that task on your own behalf?

With kind regards,
Andreas.

**Comment by jeeminso** [comment]
thank you Andreas @amotl, i will fix this

**Comment by amotl** [comment]
Thanks, @jeeminso. As pointed out by GitHub (_This branch is out-of-date with the base branch_), you might also want to rebase on top of current master while being at it.

**Comment by amotl** [comment]
Dear @jeeminso,

thanks again for updating this PR. With your recent improvements, the test harness outcome went green. Cheers!

I will leave just a minor comment about the grammar and humbly ask @mfussenegger or @seut to have a look at the overall patch again.

Apart from that, we should add respective documentation and maybe another integration test case which covers the full "CREATE TABLE AS" statement beyond the `ColumnDefinitionExtractorTest`.

So, I am humbly asking if you still feel like you want to add these improvements to this PR or would you like us to take over?

With kind regards,
Andreas.

File: libs/sql-parser/src/main/antlr/SqlBase.g4 [hunk 2]
```diff
@@ -478,6 +478,7 @@ createStmt
     : CREATE TABLE (IF NOT EXISTS)? table
         '(' tableElement (',' tableElement)* ')'
          partitionedByOrClusteredInto withProperties?                                #createTable
+    | CREATE TABLE table AS '(' query ')'                                            #createTableAs
```
Comment (amotl) [inline libs/sql-parser/src/main/antlr/SqlBase.g4:481]: The "Compatibility" section of the respective PostgreSQL documentation about "CREATE TABLE AS" says [1]:

> `CREATE TABLE AS` conforms to the SQL standard. The following are nonstandard extensions:
> - The standard requires parentheses around the subquery clause; in PostgreSQL, these parentheses are optional.

According to @seut, this can be implemented like

```suggestion
    | CREATE TABLE table AS insertSource                                             #createTableAs
```

Can you adjust this bit here?

[1] https://www.postgresql.org/docs/current/sql-createtableas.html#id-1.9.3.86.9

File: server/src/main/java/io/crate/analyze/ColumnDefinitionExtractor.java [hunk 3]
```diff
@@ -0,0 +1,259 @@
...
+    private final FieldsAstVisitor fieldsAstVisitor = new FieldsAstVisitor();
+
+
+    public ColumnDefinitionExtractor(Schemas schemas) {
+        this.schemas = schemas;
+    }
+
+    public List<TableElement<Expression>> extractFrom(AnalyzedRelation analyzedRelation){
+        return analyzedRelation.accept(queryAnalyzedRelationVisitor,null);
+    }
```
Comment (mfussenegger) [inline server/src/main/java/io/crate/analyze/ColumnDefinitionExtractor.java]: I think I don't really understand why the visiting of the relations is necessary.

Wouldn't it be sufficient to do something like this:


```java
return Lists2.map(analyzedRelation.outputs(), this::outputToColumnDefinition)
```

And `outputToColumnDefinition` would look something like this:


```java
public static ColumnDefinition<Expression> outputToColumnDefinition(Symbol symbol) {
    ColumnIdent column = Symbols.pathFromSymbol(symbol);

    var type = MetadataToASTNodeResolver.dataTypeToColumnType(...);
    return new ColumnDefinition<>(
      column.name(),
      null,       -- default expression is always null
      null,       -- generated expression is always null
      type,
      List.of()   // constraints are always empty
    )
}
```

It looks like constraints are never copied from the source, so the above should be sufficient to handle all source relations in a generic way.



PostgreSQL:

```
psql=# create table tbl (x int default 10);
Time: 16.267 ms
psql=# create table copy as (select * from tbl);
Time: 3.355 ms
psql=# \d copy
                Table "public.copy"
 Column |  Type   | Collation | Nullable | Default
--------+---------+-----------+----------+---------
 x      | integer |           |          |
```

Comment (amotl) [reply server/src/main/java/io/crate/analyze/ColumnDefinitionExtractor.java]: FWIW, the optional `like_option` clauses to the general `CREATE TABLE` statement [1] will allow users to specify which additional properties of the original table to copy.

The SO article at [2] outlines this a bit more and gives some examples.
```sql
-- Apply all properties from the source table.
CREATE TABLE IF NOT EXISTS <ClonedTableName> (LIKE <OriginalTableName> INCLUDING ALL);

-- Apply specific properties from the source table.
CREATE TABLE <ClonedTableName> (LIKE <OriginalTableName> INCLUDING DEFAULTS INCLUDING CONSTRAINTS INCLUDING INDEXES);
```

[1] https://www.postgresql.org/docs/current/sql-createtable.html
[2] https://stackoverflow.com/questions/11176383/create-table-as-with-primary-key-in-one-statement-postgresql

---

So, would the visiting of the relations as implemented by @jeeminso be suitable to reuse for implementing the `CREATE TABLE LIKE` feature instead?

File: libs/sql-parser/src/main/antlr/SqlBase.g4 (hunk 2 above)
Comment (jeeminso) [reply libs/sql-parser/src/main/antlr/SqlBase.g4:481]: @amotl Thank you Andreas, will fix this

File: server/src/main/java/io/crate/analyze/ColumnDefinitionExtractor.java (hunk 3 above)
Comment (mfussenegger) [reply server/src/main/java/io/crate/analyze/ColumnDefinitionExtractor.java]: > So, would the visiting of the relations as implemented by @jeeminso be suitable to reuse for implementing the CREATE TABLE LIKE feature instead?

I think it requires a different implementation. `LIKE` is restricted to a table name, not an arbitrary relation, so it won't be necessary to handle different concrete `AnalyzedRelation` implementations

Comment (jeeminso) [reply server/src/main/java/io/crate/analyze/ColumnDefinitionExtractor.java]: > It looks like constraints are never copied from the source, so the above should be sufficient to handle all source relations in a generic way.

Yes, if constraints are not required, outputToColumnDefinition() should be sufficient.
One question, I tried it on MySQL and observed the constraints copied, does CrateDB references Postgres documents only?

```SQL
mysql> create table tbl (x int default 10);
Query OK, 0 rows affected (0.06 sec)

mysql> show create table tbl;
+-------+----------------------------------------------------------------------------------------+
| Table | Create Table                                                                           |
+-------+----------------------------------------------------------------------------------------+
| tbl   | CREATE TABLE `tbl` (
  `x` int(11) DEFAULT '10'
) ENGINE=InnoDB DEFAULT CHARSET=latin1 |
+-------+----------------------------------------------------------------------------------------+
1 row in set (0.00 sec)

mysql> create table copy as (select * from tbl);
Query OK, 0 rows affected (0.06 sec)
Records: 0  Duplicates: 0  Warnings: 0

mysql> show create table copy;
+-------+-----------------------------------------------------------------------------------------+
| Table | Create Table                                                                            |
+-------+-----------------------------------------------------------------------------------------+
| copy  | CREATE TABLE `copy` (
  `x` int(11) DEFAULT '10'
) ENGINE=InnoDB DEFAULT CHARSET=latin1 |
+-------+-----------------------------------------------------------------------------------------+
1 row in set (0.00 sec)

```

Comment (mfussenegger) [reply server/src/main/java/io/crate/analyze/ColumnDefinitionExtractor.java]: >  does CrateDB references Postgres documents only?

yes, we try to follow the SQL Standard and if there are behaviour extensions we try to mimic PostgreSQL.

**Comment by jeeminso** [comment]
@amotl @mfussenegger Thank you for your help. If you do not mind, I would like to try to complete it myself. 

> Dear @jeeminso,
> 
> thanks again for updating this PR. With your recent improvements, the test harness outcome went green. Cheers!
> 
> I will leave just a minor comment about the grammar and humbly ask @mfussenegger or @seut to have a look at the overall patch again.
> 
> Apart from that, we should add respective documentation and maybe another integration test case which covers the full "CREATE TABLE AS" statement beyond the `ColumnDefinitionExtractorTest`.
> 
> So, I am humbly asking if you still feel like you want to add these improvements to this PR or would you like us to take over?
> 
> With kind regards,
> Andreas.

File: server/src/main/java/io/crate/analyze/ColumnDefinitionExtractor.java (hunk 3 above)
Comment (jeeminso) [reply server/src/main/java/io/crate/analyze/ColumnDefinitionExtractor.java]: thanks for clarification, will drop the 'visiting relations' logic and proceed.

Comment (amotl) [reply server/src/main/java/io/crate/analyze/ColumnDefinitionExtractor.java]: > Also: I think for a first version it's okay to **not** support the LIKE clause. We can extend that later in separate PRs. Otherwise this one will become rather big.

Sure. I didn't want to argue otherwise and just wanted to support this conversation by giving a perspective into the other syntax which supports transferring table properties. Thanks for sorting this out, @jeeminso and @mfussenegger!

**Comment by amotl** [comment]
> Thank you for your help. If you do not mind, I would like to try to complete it myself.

Sure, please go ahead!

File: server/src/main/java/io/crate/analyze/ColumnDefinitionExtractor.java (hunk 3 above)
Comment (jeeminso) [reply server/src/main/java/io/crate/analyze/ColumnDefinitionExtractor.java]: I have spent some time on this and encountered one issue. If one of the outputs is a type of ScopedSymbol with its dataType being an ObjectType, its ColumnPolicy cannot be recovered without 'visiting relations'. Please help clarify whether recovering ColumnPolicy is required. 

I have tried to find Postgres equivalent of ObjectType and it seems that CompositeType is the closest but does not have ColumnPolicy feature.

Currently, I assigned ColumnPolicy.STRICT which should be default for ObjectTypes. 

Attached below the code snippet, in case you are interested. (almost all of the changes since the last changeset)
```java
    private List<TableElement<Expression>> getTableElementsFromSourceQuery(Query sourceQuery,
                                                                           ParamTypeHints paramTypeHints,
                                                                           CoordinatorTxnCtx txnCtx) {

        AnalyzedRelation analyzedSourceQuery = relationAnalyzer.analyze(
            sourceQuery,
            new StatementAnalysisContext(paramTypeHints, Operation.READ, txnCtx));

        return Lists2.map(analyzedSourceQuery.outputs(), CreateTableAsAnalyzer::outputToColumnDefinition);
    }

    private static ColumnDefinition<Expression> outputToColumnDefinition(Symbol symbol) {

        final ColumnIdent column = Symbols.pathFromSymbol(symbol);
        final ColumnPolicy columnPolicy = ColumnPolicy.STRICT;
        final DataType<?> dataType = symbol.valueType();
        var columnType = dataTypeToColumnType(dataType, columnPolicy);

        return new ColumnDefinition<>(
            column.name(),
            null,
            null,
            columnType,
            List.of()
        );
    }

    //The original version from MetadataToASTNodeResolver depends on its private Extractor 
    // that makes use of extractColumnDefinitions() and DocTableInfo
    private static ColumnType<Expression> dataTypeToColumnType(DataType<?> type,
                                                               ColumnPolicy columnPolicy) {

        if (type.id() == ObjectType.ID) {
            List<ColumnDefinition<Expression>> childColumns = new ArrayList<>();
            ((ObjectType) type).innerTypes().forEach(
                (childName, childDataType) -> {
                    childColumns.add(
                        new ColumnDefinition<>(
                            childName,
                            null,
                            null,
                            dataTypeToColumnType(childDataType, columnPolicy),
                            List.of())
                    );
                }
            );
            return new ObjectColumnType<>(columnPolicy.name(), childColumns);
        } else if (type.id() == ArrayType.ID) {
            DataType<?> innerType = ((ArrayType<?>) type).innerType();
            return new CollectionColumnType<>(dataTypeToColumnType(
                innerType,
                columnPolicy
            ));
        } else if (type.id() == StringType.ID) {
            StringType stringType = (StringType) type;
            if (stringType.unbound()) {
                return new ColumnType<>(type.getName());
            } else {
                return new ColumnType<>("VARCHAR", List.of(stringType.lengthLimit()));
            }
        } else {
            return new ColumnType<>(type.getName());
        }
    }

```

Comment (mfussenegger) [reply server/src/main/java/io/crate/analyze/ColumnDefinitionExtractor.java]: I think for now it is okay to proceed with defaulting to `ColumnPolicy.STRICT`.

We may have to refactor the structures later to expose the column policy as part of the object-type, but we can follow up on that at a later point.

**Comment by jeeminso** [comment]
I am currently in the process of writing unit tests, please help clarify a behaviour.
```SQL
create table tbl (col object(strict) as (nested_col text));
create table cpy as select col['nested_col'] from tbl;
cr> select * from cpy;
```
From the above sql, should I get a column with the name, "nested_col" or "col['nested_col']" ?

I have opened more detailed feature request #10834 regarding this. Thank you!!

**Comment by jeeminso** [comment]
along with the question above, I have a new changeset commited:
- removed the Extractor
- fixed the grammar for brackets
- refactored CreateTableAsAnalyzer and created SymbolToColDefConverter and its unittests.
- written just a single test case written for CreateTableAsAnalyzer since it does not have much new logic.

Will look into plannerTest and IntegrationTest. Thank you.

**Comment by mfussenegger** [comment]
> From the above sql, should I get a column with the name, "nested_col" or "col['nested_col']" ?

`col['nested_col']` would currently be the right behavior

**Comment by jeeminso** [comment]
@mfussenegger Thank you for your response. I have made the change but now there is an exception thrown
```
cr> create table cpy as select col['nested_col'] from tbl;
InvalidColumnNameException["col['nested_col']" conflicts with subscript pattern]
```

File: server/src/main/java/io/crate/analyze/SymbolToColumnDefinitionConverter.java [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+import java.util.List;
+
+public class SymbolToColumnDefinitionConverter {
+
+    private static final ColumnPolicy OBJECT_TYPE_DEFAULT_COLUMN_POLICY = ColumnPolicy.STRICT;
+
+    public static ColumnDefinition<Expression> convert(Symbol symbol) {
+        return symbolToColumnDefinition(symbol);
+    }
+
```
Comment (mfussenegger) [inline server/src/main/java/io/crate/analyze/SymbolToColumnDefinitionConverter.java:49]:
```suggestion
```

I think going with just `symbolToColumnDefinition` directly should be good enough?

File: server/src/main/java/io/crate/analyze/SymbolToColumnDefinitionConverter.java [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+        return new ColumnDefinition<>(
+            columnName,
+            null,
+            null,
+            columnType,
+            List.of());
+    }
+
+    //The original version from MetadataToASTNodeResolver depends on its private Extractor
+    // that makes use of extractColumnDefinitions() and DocTableInfo
```
Comment (mfussenegger) [inline server/src/main/java/io/crate/analyze/SymbolToColumnDefinitionConverter.java:68]: Would it be possible to adapt the code in `MetadataToASTNodeResolver` to use a shared version?

**Review by mfussenegger — COMMENTED** [review]
> Thank you for your response. I have made the change but now there is an exception thrown
> 
> ```sql
> cr> create table cpy as select col['nested_col'] from tbl;
> InvalidColumnNameException["col['nested_col']" conflicts with subscript pattern]
> ```

Oh I see. I think it would be okay to have this as a "not supported" case for now. Users could still use aliases to have something like `SELECT col['nested_col'] as nested_col`. So there is at least a workaround.

Left some other minor comments. Overall looks really good already. I saw that there is some commented out code. I assume that is still in WIP?

File: server/src/main/java/io/crate/analyze/SymbolToColumnDefinitionConverter.java (hunk 5 above)
Comment (jeeminso) [reply server/src/main/java/io/crate/analyze/SymbolToColumnDefinitionConverter.java:68]: ok thanks, will look into it. 
are you referring to the commented code in the testcase? it was just for me to use as a reference, will remove that, for non-draft PR. Thank you

**Comment by jeeminso** [comment]
- commented code removed
- unnecessary wrapper method removed
- merged to the shared version of DataTypeToColumnType()

I didn't think any further testing is required, please advise.
If so, documentation should be the only task remaining by now.

File: resolv.conf [hunk 6]
```diff
@@ -0,0 +1 @@
+../run/systemd/resolve/resolv.conf
```
Comment (mfussenegger) [inline resolv.conf:1]: Looks like this file as added by accident. Could you remove it?

File: server/src/main/java/io/crate/analyze/MetadataToASTNodeResolver.java [hunk 7]
```diff
@@ -109,7 +124,7 @@ public static Expression expressionFromColumn(ColumnIdent ident) {
             if (stringType.unbound()) {
                 return new ColumnType<>(type.getName());
             } else {
-                return new ColumnType<>("VARCHAR", List.of(stringType.lengthLimit()));
+                return new ColumnType<>("varchar", List.of(stringType.lengthLimit()));
```
Comment (mfussenegger) [inline server/src/main/java/io/crate/analyze/MetadataToASTNodeResolver.java:127]: Why did this change to lowercase?

File: server/src/main/java/io/crate/analyze/MetadataToASTNodeResolver.java [hunk 8]
```diff
@@ -283,9 +313,9 @@ public Extractor(DocTableInfo tableInfo) {
             GenericProperties<Expression> properties = new GenericProperties<>();
             Expression numReplicas = new StringLiteral(tableInfo.numberOfReplicas());
             properties.add(new GenericProperty<>(
-                TableParameters.NUMBER_OF_REPLICAS.getKey(),
-                numReplicas
-                )
+                               TableParameters.NUMBER_OF_REPLICAS.getKey(),
+                               numReplicas
+                           )
```
Comment (mfussenegger) [inline server/src/main/java/io/crate/analyze/MetadataToASTNodeResolver.java:318]: Could you revert the unrelated formatting changes here?

File: server/src/main/java/io/crate/analyze/MetadataToASTNodeResolver.java [hunk 9]
```diff
@@ -294,9 +324,9 @@ public Extractor(DocTableInfo tableInfo) {
             );
             for (Map.Entry<String, Object> entry : tableParameters.entrySet()) {
                 properties.add(new GenericProperty<>(
-                        TableParameters.stripIndexPrefix(entry.getKey()),
-                        Literal.fromObject(entry.getValue())
-                    )
+                                   TableParameters.stripIndexPrefix(entry.getKey()),
+                                   Literal.fromObject(entry.getValue())
+                               )
```
Comment (mfussenegger) [inline server/src/main/java/io/crate/analyze/MetadataToASTNodeResolver.java:329]: Same here

**Review by mfussenegger — COMMENTED** [review]
Left some more minor comments.

I think only documentation and maybe a full integration level test case is missing and otherwise this should be good to get merged.
See classes deriving from `SQLTransportIntegrationTest` for examples for a full integration level test case.

**Comment by jeeminso** [comment]
- reverted the indentations and deleted resolve.conf, thanks for catching them
- I changed "VARCHAR" to "varchar", because it caused SQLParseException[Cannot find data type: VARCHAR]. When the re-created ColumnDefinitions goes through TableElementsAnalysis, TYPES_BY_NAME_OR_ALIAS (a Map in DataType class) is referenced with "varchar" as the key. I could not find elsewhere to make it lowercase.
- included a few integration tests
- started on documentation

**Comment by jeeminso** [comment]
I have updated the docs, let me know if it needs further details.
Also, please help clarify what is the purpose of ``sql_features`` table. I am not sure what modification it requires regarding this feature. Thank you

**Comment by jeeminso** [comment]
the documentation build is failing, sorry will look into it, having trouble setting up my local env

**Comment by mfussenegger** [comment]
@jeeminso I saw that the documentation tests now pass. Is there anything else missing from your side or is this ready for another round of review?

**Comment by jeeminso** [comment]
@mfussenegger it is ready for review. thank you

File: docs/appendices/release-notes/unreleased.rst [hunk 10]
```diff
@@ -53,7 +53,7 @@ None
 Changes
 =======
 
-None
+- added support for handling create-table-as statement.
```
Comment (mfussenegger) [inline docs/appendices/release-notes/unreleased.rst:56]:
```suggestion
- Added support for the :ref:`CREATE TABLE AS <ref-create-table-as>` statement.
```

**Review by mfussenegger — APPROVED** [review]
Left a suggestion for the release notes entry, otherwise this looks good to me. Nice work :+1:

**Comment by jeeminso** [comment]
@mfussenegger @amotl thank you for your help!

**Review by matthijskrul — APPROVED** [review]

**Comment by jeeminso** [comment]
should I squash all into a single commit from my end?

**Comment by mfussenegger** [comment]
> should I squash all into a single commit from my end?

No need, I rebased, squashed and merged.

Thank you again for the contribution :+1:
